### PR TITLE
Automatically split large metrics, tags, or params into smaller chunks to avoid request limit error

### LINKED ISSUE #6049: [FR] Automatically split metrics, tags, etc into smaller chunks to avoid request limit error
### Willingness to contribute

Yes. I would be willing to contribute this feature with guidance from the MLflow community.

### Proposal Summary

As in the title, MLflow might be able to avoid request limits by automatically splitting data that contains too many elements into smaller ones when we call `mlflow.log_metrics`, `mlflow.log_params`, and `mlflow.set_tags` (possible other API, too).   

### Motivation

> #### What is the use case for this feature?

Suppose `mlflow.log_metrics` case. We would like to save many parameters at once as follows:

```python
import mlflow


metrics = {f"param_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH+1)}

# Log a batch of metrics
with mlflow.start_run():
    mlflow.log_metrics(metrics)
```

However, mlflow yields the following error:

```
File /opt/homebrew/Caskroom/miniconda/base/envs/optuna/lib/python3.9/site-packages/mlflow/utils/validation.py:296, in _validate_batch_limit(entity_name, limit, length)
    290 if length > limit:
    291     error_msg = (
    292         "A batch logging request can contain at most {limit} {name}. "
    293         "Got {count} {name}. Please split up {name} across multiple requests and try "
    294         "again."
    295     ).format(name=entity_name, count=length, limit=limit)
--> 296     raise MlflowException(error_msg, error_code=INVALID_PARAMETER_VALUE)

MlflowException: A batch logging request can contain at most 1000 metrics. Got 1001 metrics. Please split up metrics across multiple requests and try again.
```

> #### Why is this use case valuable to support for MLflow users in general?

To avoid this, users need to split `metrics` into smaller subsets like 

```python
from itertools import islice
import mlflow


metrics = {f"param_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH+1)}


# Log a batch of metrics
with mlflow.start_run():
    if len(metrics) > mlflow.utils.validation.MAX_METRICS_PER_BATCH:
        it = iter(metrics)
        for _ in range(0, len(metrics), mlflow.utils.validation.MAX_METRICS_PER_BATCH):
            sub_metrics = {k: metrics[k] for k in islice(it, mlflow.utils.validation.MAX_METRICS_PER_BATCH)}
            mlflow.log_metrics(sub_metrics)
    else:
        mlflow.log_metrics(metrics)
```

I suppose this can handle in `mlflow.log_metrics`. By doing so, users do not need to care about the number of elements of `metrics`. 

> #### Why is this use case valuable to support for your project(s) or organization?

I'm from [optuna](https://github.com/optuna/optuna), a black-box optimisation framework library, community. Optuna provides an MLFlow callback that enables us to save optimisation results by using MLFlow API. If this feature request can be done by MLFlow side, we do not need the added changes by https://github.com/optuna/optuna/pull/3651.

> #### Why is it currently difficult to achieve this use case?

N/A

### Details

_No response_

### What component(s) does this bug affect?

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [X] `area/tracking`: Tracking Service, tracking client APIs, autologging

### What interface(s) does this bug affect?

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

### What language(s) does this bug affect?

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

### What integration(s) does this bug affect?

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### PR DESCRIPTION
**Pull request by nzw0301**
## Related Issues/PRs

Resolve #6049

## What changes are proposed in this pull request?

This PR aims to automatically split large reported data into smaller ones to avoid [API request limitations](https://www.mlflow.org/docs/latest/rest-api.html#request-limits).

## How is this patch tested?

- [ ] I have written tests (not required for typo or doc fix) and confirmed the proposed feature/bug-fix/change works.

Manually confirmed by running the following codes depending on the store backend.

### File

```python
import mlflow
from mlflow.tracking import MlflowClient

metrics = {f"metric_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH*2)}
with mlflow.start_run():
    mlflow.log_metrics(metrics)

mlfl_client = MlflowClient()
experiment = mlfl_client.list_experiments()[0]
run_infos = mlfl_client.list_run_infos(experiment.experiment_id)

run = mlfl_client.get_run(run_infos[0].run_id)
run_dict = run.to_dictionary()

assert len(run_dict["data"]["metrics"]) == mlflow.utils.validation.MAX_METRICS_PER_BATCH*2
```

### RDB

```python
import mlflow
from mlflow.tracking import MlflowClient

url = 'sqlite:///mlruns.db' 
mlflow.set_tracking_uri(url)

metrics = {f"metric_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH*2)}

# Log a batch of metrics
with mlflow.start_run():
    mlflow.log_metrics(metrics)

mlfl_client = MlflowClient(url)
experiment = mlfl_client.list_experiments()[0]
run_infos = mlfl_client.list_run_infos(experiment.experiment_id)

run = mlfl_client.get_run(run_infos[0].run_id)
run_dict = run.to_dictionary()
assert len(run_dict["data"]["metrics"]) == mlflow.utils.validation.MAX_METRICS_PER_BATCH*2
```

### API

```sh
mlflow server \
    --backend-store-uri sqlite:///mlruns_api.db \
    --default-artifact-root artifact --port 5001
```

```python
import mlflow
from mlflow.tracking import MlflowClient

url = 'http://localhost:5001/'
mlflow.set_tracking_uri(url)

# with split logic case
metrics = {f"metric_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH*2)}

with mlflow.start_run():
    mlflow.log_metrics(metrics)

mlfl_client = MlflowClient(url)
experiment = mlfl_client.list_experiments()[0]
run_infos = mlfl_client.list_run_infos(experiment.experiment_id)

run = mlfl_client.get_run(run_infos[0].run_id)
run_dict = run.to_dictionary()
assert len(run_dict["data"]["metrics"]) == mlflow.utils.validation.MAX_METRICS_PER_BATCH*2

## Without split logic case
metrics = {f"metric_{i}": i for i in range(mlflow.utils.validation.MAX_METRICS_PER_BATCH-10)}

with mlflow.start_run():
    mlflow.log_metrics(metrics)

mlfl_client = MlflowClient(url)
experiment = mlfl_client.list_experiments()[0]
run_infos = mlfl_client.list_run_infos(experiment.experiment_id)

run = mlfl_client.get_run(run_infos[0].run_id)
run_dict = run.to_dictionary()
assert len(run_dict["data"]["metrics"]) == mlflow.utils.validation.MAX_METRICS_PER_BATCH-10
```

## Does this PR change the documentation?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

`log_batch` can save `metrics`, `params`, and `tags` whose element sizes are more than API limitation. 

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by nzw0301** [comment]
In my understanding, `mlflow.utils.validation._validate_batch_log_limits` and `mlflow.utils.validation._validate_batch_limit` are no longer called in `mlfow` Should I remove them and related test too?

```sh
/Users/nzw/Documents/mlflow> grep -i "_validate_batch_limit" -r .
./mlflow/utils/validation.py:def _validate_batch_limit(entity_name, limit, length):
./mlflow/utils/validation.py:    _validate_batch_limit(entity_name="metrics", limit=MAX_METRICS_PER_BATCH, length=len(metrics))
./mlflow/utils/validation.py:    _validate_batch_limit(entity_name="params", limit=MAX_PARAMS_TAGS_PER_BATCH, length=len(params))
./mlflow/utils/validation.py:    _validate_batch_limit(entity_name="tags", limit=MAX_PARAMS_TAGS_PER_BATCH, length=len(tags))
./mlflow/utils/validation.py:    _validate_batch_limit(

/Users/nzw/Documents/mlflow> grep -i "_validate_batch_log_limits" -r .
./tests/utils/test_validation.py:    _validate_batch_log_limits,
./tests/utils/test_validation.py:def test_validate_batch_log_limits():
./tests/utils/test_validation.py:                _validate_batch_log_limits(**final_kwargs)
./tests/utils/test_validation.py:        _validate_batch_log_limits(too_many_metrics[:900], too_many_params[:51], too_many_tags[:50])
./tests/utils/test_validation.py:    _validate_batch_log_limits(too_many_metrics[:1000], [], [])
./tests/utils/test_validation.py:    _validate_batch_log_limits([], too_many_params[:100], [])
./tests/utils/test_validation.py:    _validate_batch_log_limits([], [], too_many_tags[:100])
```

File: mlflow/store/tracking/rest_store.py [hunk 1]
```diff
@@ -300,13 +305,41 @@ def get_experiment_by_name(self, experiment_name):
...
+            and num_tags <= MAX_PARAMS_TAGS_PER_BATCH
+        ):
+            req_body = message_to_json(
+                LogBatch(metrics=metric_protos, params=param_protos, tags=tag_protos, run_id=run_id)
+            )
+            self._call_endpoint(LogBatch, req_body)
+        else:
+            send_long_list(run_id, metric_protos, "metrics", MAX_METRICS_PER_BATCH)
+            send_long_list(run_id, param_protos, "params", MAX_PARAMS_TAGS_PER_BATCH)
+            send_long_list(run_id, tag_protos, "tags", MAX_PARAMS_TAGS_PER_BATCH)
```
Comment (dbczumar) [inline mlflow/store/tracking/rest_store.py:342]: Instead of special casing based on the batch size, can we always chunk the inputs into batches and call `log_batch` until all chunks are consumed? There's some example code for this here that we can use as inspiration: https://github.com/mlflow/mlflow/blob/be8047ceb8c258d1a598e5230564f3aecf536e3d/mlflow/utils/autologging_utils/client.py#L295-L329.

Comment (dbczumar) [reply mlflow/store/tracking/rest_store.py:342]: This example code also has the benefit of including tags, parameters, and metrics in the *same* batched request

File: mlflow/store/tracking/rest_store.py [hunk 2]
```diff
@@ -300,13 +305,41 @@ def get_experiment_by_name(self, experiment_name):
             return None
 
     def log_batch(self, run_id, metrics, params, tags):
```
Comment (dbczumar) [inline mlflow/store/tracking/rest_store.py:307]: Rather than modifying `RestStore`, can we move these changes to `MlflowClient.log_batch()`?

Comment (nzw0301) [reply mlflow/store/tracking/rest_store.py:307]: In my understanding, we cannot detect the case, where [`_validate_param_keys_unique`](https://github.com/mlflow/mlflow/blob/17797ec97d9b4be355a756bb99efb7117c1b71a4/mlflow/utils/validation.py#L211-L226) raises `MlflowException` due to duplicated key params, by chunking batch data before calling `_validate_param_keys_unique` from `log_batch` on `mlflow/store/tracking/file_store.py` and `mlflow/store/tracking/sqlalchemy_store.py`. This was why I modified `RestStore` rather than `MLflowClienet.log_batch`.

File: mlflow/store/tracking/rest_store.py (hunk 1 above)
Comment (nzw0301) [reply mlflow/store/tracking/rest_store.py]: Thank you for sharing it! It makes consistent with the existing logic. I've updated the PR by following the shared part.

File: mlflow/store/tracking/rest_store.py (hunk 2 above)
Comment (nzw0301) [reply mlflow/store/tracking/rest_store.py]: Thank you for your summary. I've moved the change in `rest_store.py` to the `MlflowClient` part.

File: mlflow/tracking/client.py [hunk 3]
```diff
@@ -915,7 +922,24 @@ def print_run_info(r):
             tags: {'t': 't'}
             status: FINISHED
         """
-        self._tracking_client.log_batch(run_id, metrics, params, tags)
```
Comment (dbczumar) [inline mlflow/tracking/client.py:918]: @nzw0301 Apologies for the confusion. `mlflow.tracking.MlflowClient` is a thin wrapper around `mlflow.tracking._tracking_service.client.TrackingServiceClient`. Can we move this implementation to `mlflow.tracking._tracking_service.client.TrackingServiceClient` instead?

**Review by dbczumar — COMMENTED** [review]
@nzw0301 Looks fantastic! Left one small comment about moving code to a slightly different location. Can we also add a small test to https://github.com/mlflow/mlflow/blob/master/tests/tracking/test_tracking.py ensuring that batches of > 1000 metrics, > 100 params, > 100 tags are logged correctly? Thank you so much for your contribution!

File: mlflow/tracking/client.py (hunk 3 above)
Comment (nzw0301) [reply mlflow/tracking/client.py:918]: Sure!

**Comment by nzw0301** [comment]
Of course. I've implemented a test for a functionality introduced by this PR. I appreciate your feedback!

File: mlflow/store/tracking/sqlalchemy_store.py [hunk 4]
```diff
@@ -1066,7 +1065,6 @@ def compute_next_token(current_size):
     def log_batch(self, run_id, metrics, params, tags):
         _validate_run_id(run_id)
         _validate_batch_log_data(metrics, params, tags)
-        _validate_batch_log_limits(metrics, params, tags)
```
Comment (dbczumar) [inline mlflow/store/tracking/sqlalchemy_store.py:1069]: I think we should still keep this validation in the backend. Can we revert this change?

File: mlflow/store/tracking/file_store.py [hunk 5]
```diff
@@ -881,7 +880,6 @@ def _overwrite_run_info(self, run_info):
     def log_batch(self, run_id, metrics, params, tags):
         _validate_run_id(run_id)
         _validate_batch_log_data(metrics, params, tags)
-        _validate_batch_log_limits(metrics, params, tags)
```
Comment (dbczumar) [inline mlflow/store/tracking/file_store.py:884]: I think we should still keep this validation in the backend. Can we revert this change?

**Review by dbczumar — COMMENTED** [review]
@nzw0301 Just a few more tiny comments. Looking forward to merging once they're addressed. Thank you for your contribution!

File: mlflow/tracking/client.py [hunk 6]
```diff
@@ -38,6 +38,7 @@
 from mlflow.utils.logging_utils import eprint
 from mlflow.utils.uri import is_databricks_uri, construct_run_url
 
+
```
[the remark this answers is not in the document]
Comment (nzw0301) [reply mlflow/tracking/client.py]: Of course, thank you for pointing out this!

**Review by dbczumar — APPROVED** [review]
LGTM! Thanks @nzw0301 !

**Comment by nzw0301** [comment]
@dbczumar My pleasure. Thank you for your constructive feedback!
